src/contracts/caching_log_query.rs

```diff
@@ -0,0 +1,214 @@
+use crate::{
+    database::{Database, DatabaseError},
+    ethereum::ProviderStack,
+};
+use async_stream::try_stream;
+use ethers::{
+    providers::{Middleware, ProviderError},
+    types::{Filter, Log, U64},
+};
+use futures::Stream;
+use serde_json::value::RawValue;
+use std::{
+    cmp::{max, min},
+    sync::Arc,
+};
+use thiserror::Error;
+
+pub struct CachingLogQuery {
+    provider:           Arc<ProviderStack>,
+    filter:             Filter,
+    page_size:          u64,
+    cache_blocks_delay: u64,
+    database:           Option<Arc<Database>>,
+}
+
+#[derive(Error, Debug)]
+pub enum Error<ProviderError> {
+    #[error(transparent)]
+    LoadLastBlock(ProviderError),
+    #[error(transparent)]
+    LoadLogs(ProviderError),
+    #[error(transparent)]
+    Database(#[from] DatabaseError),
+    #[error(transparent)]
+    Parse(serde_json::Error),
+    #[error("empty block index")]
+    EmptyBlockIndex,
+    #[error("empty transaction index")]
+    EmptyTransactionIndex,
+    #[error("empty log index")]
+    EmptyLogIndex,
+}
+
+impl CachingLogQuery {
+    pub fn new(provider: Arc<ProviderStack>, filter: &Filter) -> Self {
+        Self {
+            provider,
+            filter: filter.clone(),
+            page_size: 10000,
```

Comment:
Is this the same as the `max_log_blocks` option?

It's important that this is configurable, or even better make it dynamic (see #240)

Reply:
Yes, `page_size` is set to `max_log_blocks` in https://github.com/worldcoin/signup-sequencer/pull/255/files#diff-6f2d98d6ba19d324003887e0fbb816a48b66ac232e041924824039ac5116a729R510

10000 is the default value from `get_logs_paginated`.